File: 3219-make-lexicographically-smallest-array-by-swapping-elements/3219-make-lexicographically-smallest-array-by-swapping-elements.cpp
class Solution {
public:
    vector<int> lexicographicallySmallestArray(vector<int>& nums, int limit) {
        int n = nums.size();
        vector<pair<int, int>> help(n);
        for (int i = 0; i < n; i++)
            help[i] = {nums[i], i};
        
        sort(help.begin(), help.end(), [&](const auto& x, const auto& y) { 
            return x.first < y.first;
        });

        vector<int> res(n, 0);

        int prev = INT_MIN;
        vector<int> pos;
        for (int s = 0, e = 0; e < n;) {
            pos.push_back(help[e].second);
            prev = help[e].first;
            e++;

            if (e == n or help[e].first - prev > limit) {
                sort(pos.begin(), pos.end());
                for (auto idx : pos) 
                    res[idx] = help[s++].first;

                pos.clear();
            }
        }

        return res;
    }
};

File: 3219-make-lexicographically-smallest-array-by-swapping-elements/3219-make-lexicographically-smallest-array-by-swapping-elements.py
class Solution:
    def lexicographicallySmallestArray(self, nums: List[int], limit: int) -> List[int]:
        n = len(nums)
        help_list = [(nums[i], i) for i in range(n)]
        help_list.sort(key=lambda x: x[0])
        
        res = [0] * n
        prev = float('-inf')
        pos = []
        
        s = 0
        e = 0
        while e < n:
            pos.append(help_list[e][1])
            prev = help_list[e][0]
            e += 1
            if e == n or help_list[e][0] - prev > limit:
                pos.sort()
                for idx in pos:
                    res[idx] = help_list[s][0]
                    s += 1
                pos.clear()
        
        return res